[SPMD] Implement split_fused_optimizer to split one fused_optimizer node to two

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98785
* __->__ #98784

Several optimization passes requires the ability to split the fused_optimizer.  This PR adds the API to support the use cases.

Differential Revision: [D44806450](https://our.internmc.facebook.com/intern/diff/D44806450/)

fegin committed:
[SPMD] Implement split_fused_optimizer to split one fused_optimizer node to two

Several optimization passes requires the ability to split the fused_optimizer.  This PR adds the API to support the use cases.

Differential Revision: [D44806450](https://our.internmc.facebook.com/intern/diff/D44806450/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Implement split_fused_optimizer to split one fused_optimizer node to two"


Several optimization passes requires the ability to split the fused_optimizer.  This PR adds the API to support the use cases.

Differential Revision: [D44806450](https://our.internmc.facebook.com/intern/diff/D44806450/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -530,3 +536,338 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+    ]
+
+
+def _split_fused_adam(
```
cc @yifuwang , IIUC, this needs to happen before we re-inplace fused Adam

fegin replied:
Yes

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -530,3 +536,338 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+    gm: IterGraphModule,
+    orig_optim_block: FusedOptimizerBlock,
+    split_gradients: Set[fx.Node],
+) -> Tuple[FusedOptimizerBlock, FusedOptimizerBlock]:
```
(no need to change this PR): splitting to two seems to be too specific? We might need to discuss how to make this generic.

fegin replied:
We can recursively use this API to split the optimizer into multiple ones. If the performance is not too bad, the code may be simpler to use this API as a base API and wrap with a recursive function.

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -530,3 +536,338 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+    """
+    Split the `orig_optim_block` into two FusedOptimizerBlock. The first one
+    will be the optimizer that optimize `split_gradients`. The second one is
+    used to optimizer the remaining gradients.
```
nit: optimizer -> optimize

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -530,3 +536,338 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+    return result[0], result[1]
+
+
+def split_fused_optimizer(
```
hmm, this one doesn't seem to be plugged into `torch.distributed._spmd.compile`? I assume we need to provide a helper `gm_transformation`?

mrshenli replied:
how do we plan to construct `split_gradients`? Manually for now?

mrshenli replied:
nvm, I saw the pr on top is using this one

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -530,3 +536,338 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+        orig_optim_indices[group_idx].append(idx)
+        # Get the argument for idx-th gradient from orig_optim_args
+        for orig_arg, optim_arg in zip(orig_optim_args, optim_args[group_idx]):
+            if orig_arg:
```
what's the convention here? Should we append a None if one optimizer state does not exist or leave the corresponding list empty?

fegin replied:
We have to leave the corresponding list empty as this is the requirement from the function call. I can add a comment to explan.

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -530,3 +536,338 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+            if orig_arg:
+                optim_arg.append(orig_arg[idx])
+
+        # If argument order of step is the same as optimizer, nothing has to be
```
Why this only happens to scalar step tensor, but not other adam states? I was assuming all other states need to go through the same set of "foreach"-like ops, but step only needs an increment?

fegin replied:
No, step is a separate foreach call which is not included in fused_adam. All other states are updated inside fused_adam.

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -530,3 +536,338 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+        old_step_idx = old_step_getitem.args[1]
+        orig_step_indices[group_idx].append(old_step_idx)
+
+    if not (
```
nit:

```python
if not all(l for l in (orig_step_indices + orig_optim_indices)):
  raise ...
```

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -530,3 +536,338 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+            for name in list_names:
+                orig_list = getattr(orig_optim_block.optim, name)
+                curr_list = getattr(optim_block, name)
+                replace_flatten_output_args(orig_list[orig_idx], curr_list[curr_idx])
```
this looks pretty expensive, we are doing: for group + for grads + for states + for grads, which can be slow for large models?

fegin replied:
The current complexity is 2 (groups) * 3 (3 states) * O(grad) * O(grad). So it's `6 * O(grad^2)`. The second grad comes from `replace_flatten_output_args`. Will rewrite it as a hash based to reduce the complexity.

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -530,3 +536,338 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+    return FusedOptimizerBlock(step, optim)
+
+
+def get_all_fused_optimizer_blocks(
```
looks like this is only used in tests. If so, let's move that to tests?

mrshenli replied:
nvm, I saw the pr on top is using this one

fegin commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -530,3 +536,338 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+                torch.ops.aten._foreach_add.Scalar,
+                (step_args, 1),
+            )
+        step_block = ForeachAddBlock(step, generate_output=True)
```
fused_adam takes state_step as an argument but it does not update state_step. See https://fburl.com/j1kt6voj.

fegin commented on test/distributed/_spmd/test_transformation.py:
```diff
@@ -188,6 +193,41 @@ def train_step(model, optim, batch):
+        )
+        for _ in range(num_iters):
+            batch = torch.randn(batch_size, dim).cuda()
+            out = train_step(model, optim, batch)
```
Done

fegin committed:
Update on "[SPMD] Implement split_fused_optimizer to split one fused_optimizer node to two"


Several optimization passes requires the ability to split the fused_optimizer.  This PR adds the API to support the use cases.

Differential Revision: [D44806450](https://our.internmc.facebook.com/intern/diff/D44806450/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Implement split_fused_optimizer to split one fused_optimizer node to two"


Several optimization passes requires the ability to split the fused_optimizer.  This PR adds the API to support the use cases.

Differential Revision: [D44806450](https://our.internmc.facebook.com/intern/diff/D44806450/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Implement split_fused_optimizer to split one fused_optimizer node to two"


Several optimization passes requires the ability to split the fused_optimizer.  This PR adds the API to support the use cases.

Differential Revision: [D44806450](https://our.internmc.facebook.com/intern/diff/D44806450/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -548,3 +554,339 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+        # If argument order of step is the same as optimizer, nothing has to be
+        # done. However, it is risky to rely on this assumption so we populate
+        # the orig_step_indices.
+        old_step_output = optim_args[group_idx].state_steps[-1]
```
what's the difference between old vs orig prefix? Or should this be `orig_step_output`?

fegin replied:
Good catch. Will change it.

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -548,3 +554,339 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+    # This is not required but calling this for consistency.
+    gm.graph.eliminate_dead_code()
+
+    return result[0], result[1]
```
nit: 

```python
return *result
```

fegin replied:
Btw, it should be `tuple(result)`. `*` can cause error in this case.

mrshenli replied:
Oh I see. please ignore my comments then

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -548,3 +554,339 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+        raise ValueError("At least one split optimizer does not have input.")
+
+    output = get_output(gm.graph)
+    result: List[FusedOptimizerBlock] = []
```
nit: `results`

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -548,3 +554,339 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+            # Replace the original step output in the graph output node with
+            # the new one.
+            orig_step_output = orig_step_outputs[i]
+            replace_flatten_output_args(orig_step_output, step_output)
```
do we need to worry about there might be other nodes other then output consuming step out?

fegin replied:
I was thinking to use `fx.Node.replace_all_uses_with()`. However, this will perform 6*O(grad) tree_unflatten/tree_flatten. So I decided to use write `replace_flatten_output_args` to speed up the performance and assume only output as the user. But you are right about the case. 

Let me use `fx.Node.replace_all_uses_with()` and add a TODO to improve the performance.

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -548,3 +554,339 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+            distance += 1
+            if nodes:
+                nodes.append(None)
+            continue
+        if node.op == OP.CALL_FUNCTION and str(node.target).startswith(
```
looks like this is the same as `elif`?

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -548,3 +554,339 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+            step_node = node
+            break
+        else:
+            flatten_args, _ = tree_flatten((node.args, node.kwargs))
+            for ancestor in flatten_args:
+                if isinstance(ancestor, fx.Node):
+                    nodes.append(ancestor)
```
nit: some simplification

```python
nodes += [a for a in tree_flatten((node.args, node.kwargs))[0] if isinstance(a, fx.Node)]
```

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -548,3 +554,339 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+                updated_arg.target == operator.getitem
+            ), f"Unexpected node target {updated_arg.target}"
+            idx = cast(int, updated_arg.args[1])
+            output_copy = next(iter(updated_arg.users))
```
and this assumes the `copy_` is immediately after `getitem`? I suppose we fully control these order as we inserted those nodes and there won't be other passes be conducted in the middle of `generate_outputs` and `populate_outputs` to break these assumptions?

fegin replied:
If you mean topological order, it does not matter. We are using data dependency (users) to traverse.

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -548,3 +554,339 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+
+
+@dataclass(unsafe_hash=True)
+class ForeachAddBlock:
```
nit (can be done in followup PR): looks like the two types of blocks share the same API, maybe create a common Protocol/ABC?

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -548,3 +554,339 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+    # Optimizer is used as the output of the train_step. Therefore, we have to
+    # update the output node of the graph.
+    output_args, output_kwargs = tree_unflatten(flatten_output_args, spec)
+    gm.graph.node_set_args(output, output_args)
```
nit: a little bit inconsistent with the convention in `iter_graph_module.py`, where we directly do `output.args = output_args` there

mrshenli reviewed (approved):
stamp to unblock. Left some minor comments.

fegin commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -548,3 +554,339 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+    param_outputs: List[fx.Node] = field(default_factory=list)
+    grad_outputs: List[fx.Node] = field(default_factory=list)
+    exp_avgs_outputs: List[fx.Node] = field(default_factory=list)
+    exp_avg_sqs_outputs: List[fx.Node] = field(default_factory=list)
```
Good point. Let me add in another PR.

fegin commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -548,3 +554,339 @@ def remove_copy_from_optimizer(gm: IterGraphModule) -> None:
+                output_list.append(output_copy)
+
+        _generate_outputs(0, self.param_outputs)
+        # Do not generate gradient out list as it is not used.
```
I feel it is better to still have the field to avoid confusion.

fegin committed:
Update on "[SPMD] Implement split_fused_optimizer to split one fused_optimizer node to two"


Several optimization passes requires the ability to split the fused_optimizer.  This PR adds the API to support the use cases.

Differential Revision: [D44806450](https://our.internmc.facebook.com/intern/diff/D44806450/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Implement split_fused_optimizer to split one fused_optimizer node to two"


Several optimization passes requires the ability to split the fused_optimizer.  This PR adds the API to support the use cases.

Differential Revision: [D44806450](https://our.internmc.facebook.com/intern/diff/D44806450/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge

fegin committed:
Update on "[SPMD] Implement split_fused_optimizer to split one fused_optimizer node to two"


Several optimization passes requires the ability to split the fused_optimizer.  This PR adds the API to support the use cases.

Differential Revision: [D44806450](https://our.internmc.facebook.com/intern/diff/D44806450/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Implement split_fused_optimizer to split one fused_optimizer node to two"


Several optimization passes requires the ability to split the fused_optimizer.  This PR adds the API to support the use cases.

Differential Revision: [D44806450](https://our.internmc.facebook.com/intern/diff/D44806450/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge
